### DISCUSSION TITLE
Why am I unable to create a product via the API?

### DISCUSSION THREAD

**Discussion by benhavo** [saleor/saleor#9367] [Q&A] [answered]
I'm having difficulty using the productCreate mutation to add a product to my catalog. The following is an example of the request body I've tried:

```
mutation {
  productCreate(input: {description: "test", name: "gqltest", productType: "UHJvZHVjdFR5cGU6OQ==", price: 3.5, category: "Q2F0ZWdvcnk6MTQ=", isPublished: true}) {
    product {
      name
      id
      description
      price {
        currency
        amount
      }
    }
    errors {
      field
      message
    }
  }
}
```

I've also tried several permutations of including and excluding the inputs above. 

I created the product type and category via the dashboard, and retrieved the IDs via the API. I have also tried the following syntax with the $input defined with the key/value pairs written out above (this was pulled from the saleor-core source from the test_product.py file):

```
mutation createProduct($input: ProductCreateInput!) {
    productCreate(input: $input) {
        product {
            id
            category {
                name
            }
            description
            chargeTaxes
            taxType {
                taxCode
                description
            }
            name
            slug
            rating
            productType {
                name
            }
            attributes {
                attribute {
                    slug
                }
                values {
                    slug
                    name
                    reference
                    richText
                    boolean
                    dateTime
                    date
                    file {
                        url
                        contentType
                    }
                }
            }
        }
        errors {
            field
            code
            message
            attributes
        }
    }
}
```

In both cases, I'm receiving the following response:

```
{
    "errors": [
        {
            "message": "Expecting value: line 1 column 1 (char 0)",
            "extensions": {
                "exception": {
                    "code": "GraphQLError",
                    "stacktrace": [
                        "graphql.error.base.GraphQLError: Expecting value: line 1 column 1 (char 0)"
                    ]
                }
            }
        }
    ]
}
```

Any suggestions for how to address this? Thanks in advance!

**Comment by johnchiverton** [comment, marked as answer]
Breakdown your `productCreate `line and put it into the GraphQL Playground. The issues lie with the `price `and `isPublished `fields, neither of which are defined within `ProductCreateInput`. If you remove those fields, you'll still have an issue with the `description `field - something I've been struggling with for most of today.

If you also remove the `description `field, you should find that your product is created.

Here's your code with the `input` line broken down and the offending fields commented out:
```
mutation {
  productCreate(
    input: {
      description: "test", 
      name: "gqltest", 
      productType: "UHJvZHVjdFR5cGU6OQ==", 
      #price: 3.5, 
      category: "Q2F0ZWdvcnk6MTQ=", 
      #isPublished: true
    }
  )
  {
    product {
      name
      id
      description
      #price {
      #  currency
      #  amount
      #}
    }
    errors {
      field
      message
    }
  }
}
```
and here's the output from that, showing the description error I'm currently having issues with:

 ```
 "error": {
    "errors": [
      {
        "message": "test... is not a valid JSONString",
        "extensions": {
          "exception": {
            "code": "GraphQLError"
          }
        }
      }
    ],
    "extensions": {
      "cost": {
        "requestedQueryCost": 1,
        "maximumAvailable": 50000
      }
    }
  }
}
```

Obviously you can't create a record in the read-only GraphQL playground, but running this on your setup (with a token) should create the product. Pain in the arse without `price `and `description `though.

**Reply by benhavo** [reply]
Thanks so much! I'll also be working on the description next - will update here when I find a solution.

**Reply by johnchiverton** [reply]
You're very welcome. Total coincidence that we're working on very similar issues at the same time, and refreshing to know it's not only me having an issue!

Look forward to seeing your description solution, and I'll post likewise should I stumble upon it.

**Reply by benhavo** [reply]
@johnchiverton FYI, I created a product via the dashboard, and inspected the resulting description in the database (results below). It appears to support some pseudo rich-text formatting. I haven't investigated fully, but hopefully, if you haven't found this already, it will help you out.

```
{
    "time": 1647550434095, 
    "blocks": [
        {
            "id": "xuuUMGGsqx", 
            "data": 
                {
                    "text": "The Coolest Test Product", 
                    "level": 1
                }, 
            "type": "header"
        }, {
            "id": "OYShNkpAnI", 
            "data": {
                "text": "This is a really cool test product&nbsp;"
            }, 
            "type": "paragraph"
        }, {
            "id": "TjX0G-j-6v", 
            "data": {
                "text": "I've always loved this test product", 
                "caption": "- James Bond", 
                "alignment": "left"
            }, 
            "type": "quote"
        }
    ], 
    "version": "2.22.2"
}
```

**Reply by johnchiverton** [reply]
@benhavo I had something similar earlier, just returning a single product and examining the description field contents. I must admit I didn't break mine down though - I should have done, though. Looking at yours I can see some of your items have additional fields. I'll do some exploring and see if I can come up with anything. Thanks!

**Reply by johnchiverton** [reply]
So I broke my response, reformatted it (quotes, etc.) and put it into my `productCreate` code in the GraphQL playground:

```
productCreate(
  input: {
    productType: "UHJvZHVjdFR5cGU6MQ==", # Product Type ID
      description: {
        # time: 1647344318652,
    	blocks: [{
          # id: "pnEF5dPXXn",
          data: {
            text: "This is the product description for the product."
          },
          type: "paragraph"
    	}],
    	version: "2.22.2"
    }
  }
)
```


Unfortunately the response was the same as before:


```
"message": "Argument \"input\" has invalid value {productType: \"UHJvZHVjdFR5cGU6MQ==\", description: {blocks: [{data: {text: \"This is the product description for the Gollum Mask product.\"}, type: \"paragraph\"}], version: \"2.22.2\"}}.\nIn field \"description\": Expected type \"JSONString\", found {blocks: [{data: {text: \"This is the product description for the Gollum Mask product.\"}, type: \"paragraph\"}], version: \"2.22.2\"}.",
```

Looks like the struggle is continuing. Maybe not tonight though!

**Comment by maarcingebala** [comment]
We use Editor.js editor in the dashboard which stores description in this JSONString format. You can try to edit the description by passing it through GraphQL variables:

```graphql
mutation ProductUpdate($id: ID!, $description: JSONString) {
  productUpdate(id: $id, input: { description: $description }) {
    errors {
      field
      message
    }
    product {
      description
    }
  }
}
```

Variables:
```JSON
{
  "id": "UHJvZHVjdDo3Mg==",
  "description": "{\"blocks\":[{\"type\":\"paragraph\",\"data\":{\"text\":\"New description\"}}]}"
}
```

**Reply by tiendq** [reply]
Thank you, it works but I wish there is a link or brief explanation here https://docs.saleor.io/docs/3.x/developer/api-reference/scalars/jsonstring
